### PR TITLE
Use ‘docker.host’ java property as a way to set Docker access for Runtime unit tests

### PR DESCRIPTION
**Pull request by jonpspri**
## Description
These changes allow new scripts using docker-gradle-plugin to use Java properties to pass Docker URL information to the test, rather than inferring it based on whisk. properties.  If the ‘docker.host’ java property is not set, the original ‘whisk.properties’ logic is used as a fallback.

## Related issue and scope

- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components

- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [X] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes

- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [X] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [X] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [X] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

**Comment by csantanapr** [comment]
@jonpspri check this PR, travis is has errors.

In addition I tried this PR locally, in docker for Mac, where it's fine to run `docker run` with no host `--host specified`

I deleted the whisk.properties and try to run the test for the runtime-nodejs runtime
```
😄  $ rm $OPENWHISK_HOME/whisk.properties
~/dev/whisk/git/apache/incubator-openwhisk-runtime-nodejs (master)
🔥  $ cat OPENWHISK_HOME/whisk.properties
cat: $OPENWHISK_HOMEwhisk.properties: No such file or directory
~/dev/whisk/git/apache/incubator-openwhisk-runtime-nodejs (master)
🍰  $ ./gradlew tests:test

> Task :tests:compileTestScala
Pruning sources from previous analysis, due to incompatible CompileSetup.

> Task :tests:test

runtime.actionContainers.NodeJs6ActionContainerTests STANDARD_OUT
    [2018-05-05T20:36:50.788Z] [INFO] Slf4jLogger started

runtime.actionContainers.NodeJs6ActionContainerTests > nodejs6action should run and report an error for script not returning a json object FAILED
    java.lang.AssertionError: '/Users/csantana23/dev/whisk/git/apache/incubator-openwhisk/whisk.properties' does not exists but required

runtime.actionContainers.NodeJs6ActionContainerTests > nodejs6action should run a node script FAILED
    java.lang.NoClassDefFoundError: Could not initialize class common.WhiskProperties

runtime.actionContainers.NodeJs8ActionContainerTests STANDARD_OUT
    [2018-05-05T20:36:51.069Z] [INFO] Slf4jLogger started

runtime.actionContainers.NodeJs8ActionContainerTests > action-nodejs-v8 should run and report an error for script not returning a json object FAILED
    java.lang.NoClassDefFoundError: Could not initialize class common.WhiskProperties

3 tests completed, 3 failed
```
It looks like something still depends on whisk.properties

**Comment by csantanapr** [comment]
Ah it looks like just having one property is enough
```
echo whisk.version.name=local > $OPENWHISK_HOME/whisk.properties
```
But now fails the same as Travis
```
runtime.actionContainers.NodeJs8ActionContainerTests > action-nodejs-v8 should support async and await FAILED
    java.lang.AssertionError: assertion failed: 'docker run' did not exit with 0: (125,,unknown shorthand flag: 'p' in -p
    See 'docker --help'.

```

File: tests/src/test/scala/actionContainers/ActionContainer.scala [hunk 1]
```diff
@@ -88,12 +88,37 @@ object ActionContainer {
...
+     * information onto the docker command line.
+     */
+    val dockerUrl = Option(System.getProperty("docker.host"))
+    val hostStr = if (dockerUrl.isDefined) {
+      s" --host ${dockerUrl.get} "
     } else {
-      " "
+      val version = WhiskProperties.getProperty("whisk.version.name")
+      // Check if we are running on docker-machine env.
+      val hostStr = if (version.toLowerCase().contains("mac")) {
```
Comment (csantanapr) [inline tests/src/test/scala/actionContainers/ActionContainer.scala]: remove `val hostStr`
```
if (version.toLowerCase().contains("mac")) {
        s" --host tcp://${WhiskProperties.getMainDockerEndpoint()} "
      } else {
        " "
      }
```

Comment (jonpspri) [reply tests/src/test/scala/actionContainers/ActionContainer.scala]: Oops.  Yep, that’ll help.

**Comment by jonpspri** [comment]
Apologies for the spent time.  I should have run this through the cases for whisk.properties still being there more thoroughly.  I’ll ping you when I get a clean Travis build.

File: tests/src/test/scala/actionContainers/ActionContainer.scala [hunk 2]
```diff
@@ -88,12 +88,37 @@ object ActionContainer {
...
+     * information onto the docker command line.
+     */
+    val dockerUrl = Option(System.getProperty("docker.host"))
+    val hostStr = if (dockerUrl.isDefined) {
+      s" --host ${dockerUrl.get} "
     } else {
-      " "
+      val version = WhiskProperties.getProperty("whisk.version.name")
+      // Check if we are running on docker-machine env.
+      if (version.toLowerCase().contains("mac")) {
```
Comment (csantanapr) [inline tests/src/test/scala/actionContainers/ActionContainer.scala]: What do you think making the unit tests work with no “whisk.properties” present at all?
By adding a check if System.getenv("OPENWHISK_HOME")/WhiskProperties.WHISK_PROPS_FILE doesn’t exist.

Comment (csantanapr) [reply tests/src/test/scala/actionContainers/ActionContainer.scala]: Same thing below 
Check if file doesn’t exist and onOSX the. Use a port
Or if file exist read if local and OSX then same as today use a port

Comment (csantanapr) [reply tests/src/test/scala/actionContainers/ActionContainer.scala]: If file doesn’t exist and not OSX then don’t use a port

**Comment by jonpspri** [comment]
PG4 1647 ⌛️

File: tests/src/test/scala/actionContainers/ActionContainer.scala (hunk 2 above)
Comment (jonpspri) [reply tests/src/test/scala/actionContainers/ActionContainer.scala]: Could be.  I was thinking this morning of stringing together several options and using whichever came back first with an affirmative result:

1). The ‘docker.host’ Java property (which is what I’ll be using in multi-arch builds)
2).  The ‘DOCKER_HOST’ environment variable (which is set on docker-machine systems)
3).  The ‘WhiskProperties’ logic if ‘WhiskProperties’ exists (add a try/catch) block
4).  Just run ‘docker info’ and see if it works
5).  Fail

I can foldLeft an ordered list of anonymous functions implementing each of those to make it fairly maintainable.  Of course, it’ll take us some time to test the tests...

Comment (csantanapr) [reply tests/src/test/scala/actionContainers/ActionContainer.scala]: Yeah sounds good

Comment (jonpspri) [reply tests/src/test/scala/actionContainers/ActionContainer.scala]: Okay.  It'll take a little time to do and test.  I'll mark this WIP for now.

**Comment by csantanapr** [comment]
@jonpspri your done with WIP and ready for review?

**Comment by csantanapr** [comment]
PG2 3111 ⌛️

**Comment by csantanapr** [comment]
PG2 3111 🔵

**Comment by csantanapr** [comment]
New PG1 2889 ⌛️

File: tests/src/test/scala/actionContainers/ActionContainer.scala [hunk 3]
```diff
@@ -145,8 +169,7 @@ object ActionContainer {
 
     // ...find out its IP address...
     val (ip, port) =
-      if (WhiskProperties.getProperty("whisk.version.name") == "local" &&
-          WhiskProperties.onMacOSX()) {
+      if (System.getProperty("os.name").toLowerCase().contains("mac")) {
```
Comment (dubee) [inline tests/src/test/scala/actionContainers/ActionContainer.scala]: @csantanapr, said to check if `DOCKER_HOST` is set here to determine if `docker-machine` is being used on a Mac.

**Comment by csantanapr** [comment]
@dubee this is ready now. 
left docker-machine the way it is today unchanged, on docker-machine `DOCKER_HOST` is always defined.

**Comment by csantanapr** [comment]
PG3/2269 🏃

File: tests/src/test/scala/actionContainers/ActionContainer.scala [hunk 4]
```diff
@@ -88,14 +88,38 @@ object ActionContainer {
...
+    val dockerCmdString = dockerBin +
+      sys.props
+        .get("docker.host")
+        .orElse(sys.env.get("DOCKER_HOST"))
+        .map(" --host " + _)
+        .getOrElse("")
+    // Test here that this actually works, otherwise throw a somewhat understandable error message
+    proc(s"$dockerCmdString info").onComplete {
+      case Success((v, _, _)) if (v != 0) =>
+        throw new RuntimeException("""Unable to connect to docker host using '$d' as command string.
```
[the remark this answers is not in the document]
Comment (csantanapr) [reply tests/src/test/scala/actionContainers/ActionContainer.scala]: I guess it needs to be `$dockerCmdString`

[the remark this answers is not in the document]
Comment (dubee) [reply tests/src/test/scala/actionContainers/ActionContainer.scala]: Add `s` in front of the string as well.

```
s"""Unable to connect to docker host using '$d' as command string.
          |The docker host is determined using the Java property 'docker.host' or
          |the envirnoment variable 'DOCKER_HOST'. Please verify that one or the
          |other is set for your build/test process.""".stripMargin
```

**Pull request merged**